### PR TITLE
TMDM-10163 Before Save Process to return a warning message

### PR DESCRIPTION
**Pull request by yjli-talend**
https://jira.talendforge.org/browse/TMDM-10163
**What is the current behavior?** (You should also link to an open issue here)
 We only have info and error message type.

**What is the new behavior?**
 Add a new 'Warning' message type,
 Return ‘Warning’ and depending on the user selection either continues the Save process or abort it

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: org.talend.mdm.core/src/com/amalto/core/save/context/BeforeSaving.java [hunk 1]
```diff
@@ -34,18 +39,22 @@
...
-import javax.xml.xpath.XPathConstants;
-import javax.xml.xpath.XPathFactory;
-
 public class BeforeSaving implements DocumentSaver {
 
     private static final XPath XPATH = XPathFactory.newInstance().newXPath();
 
+    private final String TYPE_INFO = "info"; //$NON-NLS-1$
+
+    private final String TYPE_WARN = "warning"; //$NON-NLS-1$
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/save/context/BeforeSaving.java]: Rename to `TYPE_WARNING`

File: org.talend.mdm.core/src/com/amalto/core/save/context/BeforeSaving.java [hunk 2]
```diff
@@ -177,7 +200,8 @@ public boolean validateFormat(String msg) {
         if (attribute == null) {
             return false;
         }
-        if (!"info".equalsIgnoreCase(attribute.getNodeValue()) && !"error".equalsIgnoreCase(attribute.getNodeValue())) { //$NON-NLS-1$ //$NON-NLS-2$
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/save/context/BeforeSaving.java]: need to refactor, consider to add a new private method to check?

File: org.talend.mdm.core/src/com/amalto/core/save/context/DocumentSaver.java [hunk 3]
```diff
@@ -19,11 +19,17 @@
  */
 public interface DocumentSaver {
 
+    public static final int MESSAGE_TYPE_INFO = 0;
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/save/context/DocumentSaver.java]: Try to reuse the constant defined in `BeforeSaving.java`

File: org.talend.mdm.core/src/com/amalto/core/save/context/BeforeSaving.java [hunk 4]
```diff
@@ -190,4 +213,8 @@ public boolean validateFormat(String msg) {
         Node messageNode = messageNodeList.item(0);
         return messageNode.getNodeType() == Node.TEXT_NODE;
     }
+    
+    private boolean validateMessageType(String type) {
+        return (TYPE_INFO.equalsIgnoreCase(type) || TYPE_WARNING.equalsIgnoreCase(type) || TYPE_ERROR.equalsIgnoreCase(type));
```
Comment (htyin) [inline org.talend.mdm.core/src/com/amalto/core/save/context/BeforeSaving.java]: `( )` the out most can be removed

File: org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsController.java [hunk 5]
```diff
@@ -142,147 +144,19 @@ private void onSaveItem(AppEvent event) {
...
-                                    @Override
-                                    public void onSuccess(final ItemBean item) {
-                                        itemsDetailPanel.initBanner(item.getPkInfoList(), item.getDescription());
-                                    }
-                                });
-                    }
-                });
+        saveItem(model, viewBean, itemBean, isCreate, isClose, isApproveWarnBeforeSaveing, isStaging, detailToolBar,
+                itemsDetailPanel, progressBar,
+                browseRecordsService);
```
Comment (htyin) [inline org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsController.java]: Format. Need line break here?

File: org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsController.java [hunk 6]
```diff
@@ -142,147 +144,19 @@ private void onSaveItem(AppEvent event) {
...
     }
 
+
```
Comment (htyin) [inline org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsController.java]: There are two empty lines.

File: org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsController.java [hunk 7]
```diff
@@ -521,4 +395,154 @@ public void onSuccess(ItemBean result) {
...
     }
+
+    private void saveItem(final ItemNodeModel model, final ViewBean viewBean, final ItemBean itemBean, final Boolean isCreate,
+            final Boolean isClose, final Boolean isApproveWarnBeforeSaveing, final Boolean isStaging,
+            final ItemDetailToolBar detailToolBar, final ItemsDetailPanel itemsDetailPanel, final MessageBox progressBar,
+            final BrowseRecordsServiceAsync browseRecordsService) {
+        browseRecordsService.saveItem(viewBean, itemBean.getIds(),
+                (new ItemTreeHandler(model, viewBean, itemBean, ItemTreeHandlingStatus.ToSave)).serializeItem(), isCreate,
+                isApproveWarnBeforeSaveing,
+                Locale.getLanguage(), new SessionAwareAsyncCallback<ItemResult>() {
```
Comment (htyin) [inline org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsController.java]: Format

File: org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsController.java [hunk 8]
```diff
@@ -521,4 +395,154 @@ public void onSuccess(ItemBean result) {
...
+                                    }
+                                });
+                                messageBox.show();
+                                return;
+                            } else {
+                                messageBox.show();
+                                setTimeout(messageBox, 1000);
+                            }
+
+                            if (message == null || "".equals(messageBox.getMessage())) { //$NON-NLS-1$
```
Comment (htyin) [inline org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsController.java]: message.isEmpty()

File: org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsController.java [hunk 9]
```diff
@@ -521,4 +395,154 @@ public void onSuccess(ItemBean result) {
...
+                        // only if save succeeds, update item bean's id and lastUpdateTime
+                        if (result.getReturnValue() != null) {
+                            itemBean.setIds(result.getReturnValue());
+                            itemBean.setLastUpdateTime(result);
+                        }
+                        progressBar.close();
+                        MessageBox messageBox = MessageUtil.generateMessageBox(result);
+                        String message = messageBox.getMessage();
+                        if (result.getStatus() == ItemResult.FAILURE) {
+                            if (message == null || "".equals(messageBox.getMessage())) { //$NON-NLS-1$
```
Comment (htyin) [inline org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/mvc/BrowseRecordsController.java]: message.isEmpty()

File: org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/BrowseRecordsServiceAsync.java [hunk 10]
```diff
@@ -100,9 +98,12 @@ void getItemNodeModel(ItemBean item, EntityModel entity, boolean isStaging, Stri
 
     void getMandatoryFieldList(String tableName, AsyncCallback<List<String>> callback);
 
-    void saveItem(ViewBean viewBean, String ids, String xml, boolean isCreate, String language, AsyncCallback<ItemResult> callback);
+    void saveItem(ViewBean viewBean, String ids, String xml, boolean isCreate, boolean isApproveWarnBeforeSaveing,
+            String language,
+            AsyncCallback<ItemResult> callback);
```
Comment (htyin) [inline org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/BrowseRecordsServiceAsync.java]: Format

File: org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/widget/ForeignKeyRowEditor.java [hunk 11]
```diff
@@ -113,14 +113,12 @@ public void onSuccess(ItemResult result) {
...
-                                        String msg = MultilanguageMessageParser.pickOutISOMessage(result.getDescription());
                                         if (result.getStatus() == ItemResult.FAILURE) {
-                                            msgBox.setTitle(MessagesFactory.getMessages().error_title());
-                                            msgBox.setButtons(MessageBox.OK);
-                                            msgBox.setIcon(MessageBox.ERROR);
-                                            msgBox.setMessage(msg == null || msg.isEmpty() ? MessagesFactory.getMessages()
-                                                    .output_report_null() : msg);
+                                            MessageBox msgBox = MessageUtil.generateMessageBox(result);
+                                            String message = msgBox.getMessage();
+                                            if (message == null || "".equals(message)) { //$NON-NLS-1$
```
Comment (htyin) [inline org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/widget/ForeignKeyRowEditor.java]: message.isEmpty()

File: org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/widget/ItemsListPanel.java [hunk 12]
```diff
@@ -567,14 +568,12 @@ public void onSuccess(ItemResult result) {
...
-                                String msg = MultilanguageMessageParser.pickOutISOMessage(result.getDescription());
+                                final MessageBox msgBox = MessageUtil.generateMessageBox(result);
+                                String message = msgBox.getMessage();
                                 if (result.getStatus() == ItemResult.FAILURE) {
-                                    msgBox.setTitle(MessagesFactory.getMessages().error_title());
-                                    msgBox.setButtons(MessageBox.OK);
-                                    msgBox.setIcon(MessageBox.ERROR);
-                                    msgBox.setMessage(msg == null || msg.isEmpty() ? MessagesFactory.getMessages()
-                                            .output_report_null() : msg);
+                                    if (message == null || "".equals(message)) { //$NON-NLS-1$
```
Comment (htyin) [inline org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/widget/ItemsListPanel.java]: message.isEmpty()

File: org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/widget/treedetail/ForeignKeyUtil.java [hunk 13]
```diff
@@ -107,6 +108,7 @@ private static void saveItem(ItemNodeModel model, final boolean isCreateForeignK
             BrowseRecordsServiceAsync service = (BrowseRecordsServiceAsync) Registry.get(BrowseRecords.BROWSERECORDS_SERVICE);
             service.saveItem(viewBean, itemBean.getIds(),
                     (new ItemTreeHandler(model, viewBean, ItemTreeHandlingStatus.ToSave)).serializeItem(), isCreate,
+                    isApproveWarnBeforeSaveing,
                     Locale.getLanguage(), new SessionAwareAsyncCallback<ItemResult>() {
```
Comment (htyin) [inline org.talend.mdm.webapp.browserecords/src/main/java/org/talend/mdm/webapp/browserecords/client/widget/treedetail/ForeignKeyUtil.java]: Format

File: org.talend.mdm.core/src/com/amalto/core/save/context/BeforeSaving.java [hunk 14]
```diff
@@ -190,4 +213,8 @@ public boolean validateFormat(String msg) {
         Node messageNode = messageNodeList.item(0);
         return messageNode.getNodeType() == Node.TEXT_NODE;
     }
+    
+    private boolean isValidatedMessageType(String type) {
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/save/context/BeforeSaving.java]: better naming of `isValidMessageType`

File: org.talend.mdm.core/src/com/amalto/core/webservice/WSPutItemWithReport.java [hunk 15]
```diff
@@ -16,6 +16,9 @@
     protected com.amalto.core.webservice.WSPutItem wsPutItem;
     protected java.lang.String source;
     protected java.lang.Boolean invokeBeforeSaving;
+    protected java.lang.Boolean warningApprovedBeforeSave = false;
+    protected String messageType;
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/webservice/WSPutItemWithReport.java]: Use `java.lang.String`  same as others?

**Pull request merged**